### DISCUSSION TITLE
Email::Sender::Transport ::SMTP not connecting to localhost on port 25

### DISCUSSION THREAD

**Discussion by farokh-ny** [postfixadmin/postfixadmin#648] [Q&A]
Howdy all.
I'm trying to set up a postfix/dovecot/postfixadmin/vacation.pl system on Ubuntu 22.04
I've got everything working except that the vacation.pl script can't (or won't) connect to localhost to send out auto replies. I'm at a loss as to why, and I've tried searching without any success.
The log_level is set to 2, and log_to_file is 1, writing to /var/log/vacation.log.
I set up a temporary domain and mailbox, and when I send email to it from outside (gmail), the message comes through, and the auto reply is triggered, but all I get in the log is below. Any pointers on what to look for, or how I can debug this? I was able to use telnet to connect to port 25 on localhost and send an email using SMTP commands, so I know that's working. A couple of things I did was to 1) set the debug flag to 1 in the smtp_params array and 2) I commented out the ssl_options (I'm using a self signed cert for the mail server and even though smtp_ssl is 0 and the port is 25, I wanted to make sure it wasn't trying to verify the ssl cert). Thanks, in advance!

2022/08/10 15:43:53 DEBUG> /usr/lib/postfixadmin/vacation.pl:449 main::send_vacation_email - Will send vacation response for <d3486718-583f-761c-ff99-2a089fc69c8f@gmail.com>: FROM: pmlaw@mon.besttechsvc.com (orig_to: pmlaw@mon.besttechsvc.com), TO: bts.farokh@gmail.com; VACATION SUBJECT: Out of Office ; VACATION BODY: I will be away from <date> until <date>.
For urgent matters you can contact <contact person>.
2022/08/10 15:43:53 ERROR> /usr/lib/postfixadmin/vacation.pl:533 main::finally {...}  - Failed to send vacation response to bts.farokh@gmail.com from pmlaw@mon.besttechsvc.com subject Out of Office: unable to establish SMTP connection to (localhost) port 25

Trace begun at /usr/share/perl5/Email/Sender/Transport/SMTP.pm line 232
Email::Sender::Transport::SMTP::_throw('Email::Sender::Transport::SMTP=HASH(0x5569e10d66c0)', 'unable to establish SMTP connection to (localhost) port 25') called at /usr/share/perl5/Email/Sender/Transport/SMTP.pm line 175
Email::Sender::Transport::SMTP::_smtp_client('Email::Sender::Transport::SMTP=HASH(0x5569e10d66c0)') called at /usr/share/perl5/Email/Sender/Transport/SMTP.pm line 241
Email::Sender::Transport::SMTP::send_email('Email::Sender::Transport::SMTP=HASH(0x5569e10d66c0)', 'Email::Abstract=ARRAY(0x5569e1223aa8)', 'HASH(0x5569e1223be0)') called at /usr/share/perl5/Email/Sender/Role/CommonSending.pm line 45
Email::Sender::Role::CommonSending::try {...}  at /usr/share/perl5/Try/Tiny.pm line 102
eval {...} at /usr/share/perl5/Try/Tiny.pm line 93
Try::Tiny::try('CODE(0x5569e12344e0)', 'Try::Tiny::Catch=REF(0x5569e1208068)') called at /usr/share/perl5/Email/Sender/Role/CommonSending.pm line 58
Email::Sender::Role::CommonSending::send('Email::Sender::Transport::SMTP=HASH(0x5569e10d66c0)', 'Email::Abstract=ARRAY(0x5569e1223aa8)', 'HASH(0x5569e12343d8)') called at /usr/share/perl5/Email/Sender/Simple.pm line 119
Email::Sender::Simple::send_email('Email::Sender::Simple', 'Email::Abstract=ARRAY(0x5569e1223aa8)', 'HASH(0x5569e1223c10)') called at /usr/share/perl5/Email/Sender/Role/CommonSending.pm line 45
Email::Sender::Role::CommonSending::try {...}  at /usr/share/perl5/Try/Tiny.pm line 102
eval {...} at /usr/share/perl5/Try/Tiny.pm line 93
Try::Tiny::try('CODE(0x5569e1223bb0)', 'Try::Tiny::Catch=REF(0x5569e0b186f8)') called at /usr/share/perl5/Email/Sender/Role/CommonSending.pm line 58
Email::Sender::Role::CommonSending::send('Email::Sender::Simple', 'Email::Simple=HASH(0x5569e11eb458)', 'HASH(0x5569e1223ce8)') called at /usr/share/perl5/Sub/Exporter/Util.pm line 57
Sub::Exporter::Util::__ANON__('Email::Simple=HASH(0x5569e11eb458)', 'HASH(0x5569e1223ce8)') called at /usr/lib/postfixadmin/vacation.pl line 530
main::try {...}  at /usr/share/perl5/Try/Tiny.pm line 102
eval {...} at /usr/share/perl5/Try/Tiny.pm line 93
Try::Tiny::try('CODE(0x5569e12236e8)', 'Try::Tiny::Finally=REF(0x5569e03548a0)') called at /usr/lib/postfixadmin/vacation.pl line 538

**Comment by DavidGoodwin** [comment]
is postfix really listening on port 25 on localhost?

**Reply by farokh-ny** [reply]
Yes:
root@linuxmail2:/usr/lib/postfixadmin# telnet localhost 25
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
220 linuxmail2 ESMTP Postfix (Ubuntu)
EHLO localhost.localdomain
250-linuxmail2
250-PIPELINING
250-SIZE 10240000
250-VRFY
250-ETRN
250-STARTTLS
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-DSN
250-SMTPUTF8
250 CHUNKING
MAIL FROM:<pmlaw@mon.besttechsvc.com>
250 2.1.0 Ok
RCPT TO:<bts.farokh@gmail.com> NOTIFY=success,failure
250 2.1.5 Ok
DATA
354 End data with <CR><LF>.<CR><LF>
Subject: Test

Testing.
.
250 2.0.0 Ok: queued as 91EF980A7D
QUIT
221 2.0.0 Bye
Connection closed by foreign host.
root@linuxmail2:/usr/lib/postfixadmin#

**Comment by farokh-ny** [comment]
I was able to get it to work by changing
        try {
            sendmail($email, { transport => $transport });
        } finally {

to
        try {
            sendmail($email);
        } finally {

When I have some time, I'm going to try and figure out why using Email::Sender::Transport ::SMTP doesn't work for me, but again, any pointers on how to do that would be helpful. I'm definitely not a perl expert by any stretch.

**Reply by DavidGoodwin** [reply]
@farokh-ny thanks for the followup. I don't have any suggestions, sorry.....

**Reply by puck** [reply]
Just to echo that I've had the same issue, switching to `sendmail($email)` also resolved it for me.

Doing a bit of digging I identified that this is due to $localaddr being set to 'localhost'. Which, totally makes sense as this is the IP address that Email::Sender::Transport::SMTP will attempt to connect from to the server. It can't bind to localhost when trying to connect to a routable IP address. If I set this to another IP address on my email server then vacation.pl runs okay.

I'll submit a PR to make this setting optional.
